### PR TITLE
Add Task: Rolling Update a DaemonSet

### PR DESCRIPTION
**Pull request by janetkuo**
@kubernetes/sig-apps-pr-reviews @kargakis @lukaszo @erictune

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2878)

### CODE REVIEW

**Comment by lukaszo** [comment]
Wow, welly well written. 

/approve

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  will be created automatically, in a controlled fashion. 
+
+## Limitations 
+
+* DaemonSet rollout history is not supported yet. 
+* DaemonSet rollback is not directly supported in `kubectl` yet. You can rollback
+  by updating DaemonSet template to match the previous version.
+* Once the template of a `RollingUpdate` DaemonSet is modified, a rolling update
+  will be triggered. All existing DaemonSet pods will be killed, even when the
+  pods are created from the same DaemonSet template. This causes restarts.
```
Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: I am not sure I understand this..

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+```
+
+Once you've found those nodes, delete some non-DaemonSet pods from the node to
+make room for new DaemonSet pods. Note that this will cause service disruption 
+if the deleted pods are not controlled by any controllers, or if the pods aren't
+replicated. This doesn't respect 
+[PodDisruptionBudget](/docs/tasks/configure-pod-container/configure-pod-disruption-budget/)
+either.
+
+#### Broken rollout
```
Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Worth mentioning clock skew

File: docs/admin/daemons.md [hunk 3]
```diff
@@ -121,7 +121,7 @@ the new DaemonSet with the different template will recognize all the existing po
 matching labels.  It will not modify or delete them despite a mismatch in the pod template.
 You will need to force new pod creation by deleting the pod or deleting the node.
 
-You cannot update a DaemonSet.
+Starting from Kubernetes version 1.6, you can [rolling update a DaemonSet](/docs/tasks/manage-daemon/update-daemon-set/).
```
Comment (devin-donnelly) [inline docs/admin/daemons.md]: "you can [perform a rolling update] on a DaemonSet"

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
+---
+assignees:
+- janetkuo
+title: Rolling Updating a DaemonSet
```
Comment (devin-donnelly) [inline docs/tasks/manage-daemon/update-daemon-set.md]: "Performing a Rolling Update on a DaemonSet"

File: docs/tasks/index.md [hunk 5]
```diff
@@ -63,6 +63,9 @@ single thing, typically by giving a short sequence of steps.
 * [Debugging a StatefulSet](/docs/tasks/manage-stateful-set/debugging-a-statefulset/)
 * [Force Deleting StatefulSet Pods](/docs/tasks/manage-stateful-set/delete-pods/)
 
+#### Managing Cluster Daemons
+* [Rolling Updating a DaemonSet](/docs/tasks/manage-daemon/update-daemon-set/)
```
Comment (devin-donnelly) [inline docs/tasks/index.md]: "Performing a Rolling Update on a DaemonSet"

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+* RollingUpdate: With `RollingUpdate` update strategy, after you update a
+  DaemonSet template, old DaemonSet pods will be killed, and new DaemonSet pods
+  will be created automatically, in a controlled fashion. 
+
+## Limitations 
+
+* DaemonSet rollout history is not supported yet. 
+* DaemonSet rollback is not directly supported in `kubectl` yet. You can rollback
+  by updating DaemonSet template to match the previous version.
+* Once the template of a `RollingUpdate` DaemonSet is modified, a rolling update
```
Comment (devin-donnelly) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Passive voice.

"If you modify the `RollingUpdate` template for a DaemonSet, Kubernetes will trigger a rolling update for that DaemonSet. All existing DaemonSet pods will be killed, even when the pods are created from the same DaemonSet template."

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 1 above)
Comment (devin-donnelly) [inline docs/tasks/manage-daemon/update-daemon-set.md]: "This causes restarts." Of what? What restarts? The pods?

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+```shell
+kubectl edit ds/<daemonset-name>
+```
+
+```shell
+kubectl patch ds/<daemonset-name> -p=<strategic-merge-patch>
+```
+
+##### Update only the container image
```
Comment (devin-donnelly) [inline docs/tasks/manage-daemon/update-daemon-set.md]: "Updating only the container image"

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 2 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Added a section below

File: docs/admin/daemons.md (hunk 3 above)
Comment (janetkuo) [inline docs/admin/daemons.md]: Done

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 4 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Done

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 6 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Done and added more explanation

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 1 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Explained

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 7 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Done

File: docs/tasks/index.md (hunk 5 above)
Comment (janetkuo) [inline docs/tasks/index.md]: Done

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 1 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Explained

**Comment by janetkuo** [comment]
@devin-donnelly @kargakis thanks for the review, comments addressed, ptal

**Comment by devin-donnelly** [comment]
Awesome, thanks @janetkuo . @kargakis , I'll let you give the tech LGTM.

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+* RollingUpdate: With `RollingUpdate` update strategy, after you update a
+  DaemonSet template, old DaemonSet pods will be killed, and new DaemonSet pods
+  will be created automatically, in a controlled fashion. 
+
+## Limitations 
+
+* DaemonSet rollout history is not supported yet. 
+* DaemonSet rollback is not directly supported in `kubectl` yet. You can rollback
+  by updating DaemonSet template to match the previous version.
+* Every time you modify the `RollingUpdate` template for a DaemonSet, Kubernetes
```
Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Is this true also for changing maxUnavailable or is it the strategy type only?

File: docs/tasks/manage-daemon/update-daemon-set.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+  will trigger a rolling update for that DaemonSet. The rolling update will kill
+  all existing DaemonSet pods and create new DaemonSet pods. Note that existing
+  DaemonSet pods will be killed regardless of their pod spec, even when they are
+  created from the same template as current DaemonSet template, e.g. when you
+  quickly update a DaemonSet template and changed it back before the rollout is
```
Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Do we have an issue to fix this?

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 8 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: `maxUnavailable` and strategy type aren't part of DaemonSet template. Is it confusing? Should I rephrase it to something like "... modify the template for a `RollingUpdate` DaemonSet"?

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 9 above)
Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: This will be fixed once we implement DaemonSet history.

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 8 above)
Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: "Every time you modify the PodTemplateSpec of a DaemonSet that is using the RollingUpgrade strategy..." would be better but this has got me thinking if we really need to call this problem out. We have been doing this in OpenShift Deployments (every new version has its own pods) and nobody has ever complained about it.

Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Not saying that we shouldn't optimize this to adopt the older pods but we probably shouldn't list it as a limitation.

Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Yeah I agree, I'll just remove this bullet point.

**Comment by janetkuo** [comment]
@kargakis ptal

File: docs/tasks/manage-daemon/update-daemon-set.md (hunk 9 above)
Comment (lukaszo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: I don't see how to do it without using hash for podtemplate. History will only help with rollback.

Comment (janetkuo) [inline docs/tasks/manage-daemon/update-daemon-set.md]: It's doable if we can map a given templateGeneration to a history

Comment (kargakis) [inline docs/tasks/manage-daemon/update-daemon-set.md]: Yes, simply using history won't solve this, we need to collapse versions in duplicate PodTemplates appropriately.

**Comment by janetkuo** [comment]
@devin-donnelly this is ready to be merged

**Comment by janetkuo** [comment]
Just squashed all commits into one
